Yum: add support for downloadonly

rm-you opened the pull request:
##### SUMMARY
Add support for the yum command line option `downloadonly`.

Some operators need to use this feature of yum, and without it are unable to use the yum module for some operations. The workaround of using the shell module to call yum is valid but very suboptimal.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
yum

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (yum_mods f0aff087b2) last updated 2018/07/12 14:48:52 (GMT -700)
  config file = None
  configured module search path = [u'/home/rm_you/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rm_you/ansible/lib/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Apr 11 2018, 07:36:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```


##### ADDITIONAL INFORMATION
For example, to do a quick test of the download_only flag:
```
- hosts: localhost
  tasks:
    - name: download mariadb package
      yum:
        name: mariadb
        state: latest
        download_only: true
      become: true
```
The output of this in -vvv mode:
```
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "allow_downgrade": false,
            "bugfix": false,
            "conf_file": null,
            "disable_excludes": "all",
            "disable_gpg_check": false,
            "disable_plugin": [],
            "disablerepo": null,
            "download_only": true,
            "enable_plugin": [],
            "enablerepo": null,
            "exclude": null,
            "install_repoquery": true,
            "installroot": "/",
            "list": null,
            "name": [
                "mariadb"
            ],
            "security": false,
            "skip_broken": false,
            "state": "latest",
            "update_cache": false,
            "update_only": false,
            "validate_certs": true
        }
    },
    "msg": "",
    "rc": 0,
    "results": [
        "Loaded plugins: changelog, fastestmirror, versionlock\nLoading mirror speeds from cached hostfile\nResolving Dependencies\n--> Running transaction check\n---> Package mariadb.x86_64 1:5.5.56-2.el7 will be installed\n--> Finished Dependency Resolution\n\nDependencies Resolved\n\n================================================================================\n Package       Arch         Version                Repository              Size\n================================================================================\nInstalling:\n mariadb       x86_64       1:5.5.56-2.el7         centos-base-prod       8.7 M\n\nTransaction Summary\n================================================================================\nInstall  1 Package\n\nTotal size: 8.7 M\nInstalled size: 49 M\nBackground downloading packages, then exiting:\nexiting because \"Download Only\" specified\n"
    ]
}
```

The use of an underscore in this flag is a style choice that I wasn't sure about -- it seems like there's a mix between using underscores and not, even though yum does not use them for most of these things. I'm open to doing it either way.

rm-you committed:
Yum: Add support for --downloadonly

rm-you commented:
Ah, I also need to add to the docs section -- I will do this momentarily.

rm-you commented:
Interesting, reading through that issue and old PR gives me some context. I'm guessing some of the same comments are going to apply? I'm definitely willing to update this to include more testing and possibly expand to include more options.
I do wonder whether checking the yum version for support is actually something that matters, given the control people have over their environments, but I haven't been able to read through everything yet as I'm on mobile. Will withhold further comment until I hear from some reviewers. :)

rm-you commented:
Thanks @Akasurde for the reference!

achinthagunasekara commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -146,6 +146,22 @@
         The disabled plugins will not persist beyond the transaction.
     required: false
     version_added: "2.5"
+  download_only:
```
Can you please organise these to preserve alphabetical ordering?

rm-you replied:
Sure thing! I was just in "append to the end" mode, but that makes sense.

rm-you replied:
Err, actually, looking closely at this again, it seems like *nothing here* is in alphabetical order... Do you just mean these two new options? If so, I guess I can, but it seems a little bit unnecessary unless you're asking me to reorganize the whole thing.

achinthagunasekara replied:
Sorry I didn't look at the reset of the items here. Usually the pattern I've seen across other items is to have it organised in alphabetical order.

Akasurde replied:
`ansible-doc` organizes arguments in alphabetical order so I don't think it will necessary to do here.

Akasurde commented:
@rm-you Take a look at integration test located here - https://github.com/ansible/ansible/blob/devel/test/integration/targets/yum/tasks/yum.yml

rm-you commented:
@akasurde Ah, that's where it is! Thanks for the pointer. I'll see if I can get something done for the testing tomorrow. Any other suggestions / issues you see that would block this from merging that I can get a head start on?

Akasurde commented:
@rm-you Cool. No issue as such which will block you for now.

maxamillion commented:
The change looks good, my only suggestion would also be tests. 👍

rm-you commented:
Changed the docstrings to the way you had them @Akasurde because I liked your formatting better, and added includes as well since your patch was doing both (and it makes sense to add them together).

Gave a first pass at tests for download_only, but I am not 100% sure about how to test disable includes/excludes. 🤔

rm-you commented:
Got a prelim test for the disableexcludes function, and it seems to show that it *isn't* working. So, I'll be figuring that out.

rm-you commented:
Augh... The python yum module's http://yum.baseurl.org/api/yum/yum/packageSack.html#packagesack and `matchPackageNames` have no way that I see to respect disableexcludes, so it bails early when it can't match the package name. :(
I got it to work by just blindly skipping that matching step when there are disableexcludes defined, but this seems really wonky/lame. Anyone have any ideas? I'll try to keep investigating, but I think either I'm missing something or else the yum module would need to be patched to support this, and relying on that would be a PITA (we'd be able to use this functionality safely some time in like 2025 given the centos release schedule)...

lunarfs commented:
I need the disableexcludes and have been working on patching the yum.py for our local ansible, i found that adding 
    if disable_excludes != None:
        my.conf.disable_excludes = disable_excludes
to yum_base() solves the issue with the base yum commands not respecting disableexcludes, ofcourse the yum_base then needs params to be able to set this.. i will try to get this working... and return with my results

lunarfs commented:
@rm-you , i did create a new PR for this as i have only ben looking at implementing disable_excludes. you will finde it here
https://github.com/ansible/ansible/pull/42510/files

maxamillion commented:
@lunarfs your PR looks good (#42510), is there any desire to try and merge these two together? cc @rm-you

rm-you commented:
Yes, absolutely. Sorry, I just got back from some PTO, but I will look at ditching my disableexcludes patch and just make this PR for the downloadonly bits.

lunarfs commented:
@maxamillion do you want a branch with bot fixes in one, or is it ok with two seperate PR's? I will squash my commits once I am sure that no further fixes is needed.

rm-you commented:
@lunarfs @maxamillion They're really two separate changes, so I imagine it should be fine to merge them as two PRs. Whoever merges second can rebase but it should be a simple one.

maxamillion commented:
Sounds good to me, two PRs it is. 👍

maxamillion commented:
shipit

maxamillion commented:
rebuild_merge

maxamillion commented:
rebuild_merge

rm-you commented:
recheck

rm-you commented:
rebuild

rm-you commented:
Not sure what I need to do to get it to acknowledge that I did update it and it should remove the needs_revision tag... @maxamillion
